src/core/io/src/4C_io_input_line.hpp

```diff
@@ -0,0 +1,633 @@
+    {
+      component_ = other.component_->clone();
+      data_ = other.data_;
+      return *this;
+    }
+
+    Component(Component&&) noexcept = default;
+    Component& operator=(Component&&) noexcept = default;
+
+    void parse_and_store_value(ValueParser& parser, InputParameterContainer& container) const
+    {
+      component_->parse_and_store_value(parser, container);
+    }
+
+    void set_default_value(InputParameterContainer& container) const
+    {
+      component_->set_default_value(container);
+    }
+
+    void print(std::ostream& stream, const InputParameterContainer& container) const
+    {
+      if (!required()) stream << "[";
+      stream << data_.name;
+      component_->print_value(stream, container);
+      if (!required()) stream << "]";
+      stream << " ";
+    }
+
+    [[nodiscard]] const std::string& name() const { return data_.name; }
+
+    [[nodiscard]] const std::string& description() const { return data_.description; }
+
+    [[nodiscard]] bool required() const { return data_.required; }
+
+
+   private:
+    /**
+     * Store a reduced version of the common data entries.
+     */
+    struct
+    {
+      /**
+       * The name or key under which the value is found in the input. This name is also used to
+       * store the value in the container.
+       */
+      std::string name;
+
+      /**
+       * An optional description of the value.
+       */
+      std::string description{};
+
+      /**
+       * Whether the value is required or optional.
+       */
+      bool required{true};
+    } data_;
+
+    //! Pointer to type-erased implementation.
+    std::unique_ptr<EraseComponentInterface> component_;
+  };
+
+  template <typename DataTypeIn>
+  struct BasicComponent
+  {
+    using DataType = std::decay_t<DataTypeIn>;
+    using StoredType = typename DataType::StoredType;
+    DataType data;
+    void parse_and_store_value(ValueParser& parser, InputParameterContainer& container) const;
+  };
+
+
+  template <typename DataTypeIn>
+  struct UserDefinedComponent
+  {
+    using DataType = std::decay_t<DataTypeIn>;
+    using StoredType = typename DataType::StoredType;
+    DataType data;
+    std::function<void(ValueParser&, InputParameterContainer&)> parse_and_store_value;
+    std::function<void(std::ostream&, const InputParameterContainer&)> print_value;
+  };
+
+  /**
+   * Helper functions to create components for an InputLine.
+   */
+  namespace InputLineBuilders
+  {
+    template <typename StoredTypeIn>
+    struct ScalarComponentData
+    {
+      using StoredType = StoredTypeIn;
+      /**
+       * The name or key under which the value is found in the input. This name is also used to
+       * store the value in the container.
+       */
+      std::string name;
+
+      /**
+       * An optional description of the value.
+       */
+      std::string description{};
+
+      /**
+       * Whether the value is required or optional.
+       */
+      std::optional<bool> required{};
```

Comment:
why does it need to be optional?

Reply:
Good catch! As I just discussed with @gilrrei I overengineered this part :) This will be dropped, as it's enough to pass a `default_value` which must mean the entry is optional. If an entry is required, there cannot be a default.